pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp

```diff
@@ -453,42 +480,73 @@ sample(const std::vector<node_type>& node_types,
+                                            k[0])];  // initialize empty vector
         }
       }
-      for (const auto& k : node_types) {
-        slice_dict[k] = {slice_dict.at(k).second,
-                         sampled_nodes_dict.at(k).size()};
-        num_sampled_nodes_per_hop_map.at(k).push_back(slice_dict.at(k).second -
-                                                      slice_dict.at(k).first);
+      at::parallel_for(
+          0, threads_edge_types.size(), 1, [&](size_t _s, size_t _e) {
+            for (auto j = _s; j < _e; j++) {
+              for (const auto& k : threads_edge_types[j]) {
+                const auto src = !csc ? std::get<0>(k) : std::get<2>(k);
+                const auto dst = !csc ? std::get<2>(k) : std::get<0>(k);
+                const auto count = num_neighbors_dict.at(to_rel_type(k))[ell];
+                auto& src_sampled_nodes = sampled_nodes_dict.at(src);
+                auto& dst_sampled_nodes = parallel
+                                              ? dst_sampled_nodes_dict.at(dst)
+                                              : sampled_nodes_dict.at(dst);
+                auto& dst_mapper = mapper_dict.at(dst);
+                auto& sampler = sampler_dict.at(k);
+                size_t begin, end;
+                std::tie(begin, end) = slice_dict.at(src);
+
+                sampler.num_sampled_edges_per_hop.push_back(0);
+
+                if (!time_dict.has_value() ||
+                    !time_dict.value().contains(dst)) {
+                  for (size_t i = begin; i < end; ++i) {
+                    sampler.uniform_sample(
+                        /*global_src_node=*/src_sampled_nodes[i],
+                        /*local_src_node=*/i, count, dst_mapper, generator,
+                        dst_sampled_nodes);
+                  }
+                } else if constexpr (!std::is_scalar<
+                                         node_t>::value) {  // Temporal:
+                  const at::Tensor& dst_time = time_dict.value().at(dst);
+                  const auto dst_time_data = dst_time.data_ptr<temporal_t>();
+                  for (size_t i = begin; i < end; ++i) {
+                    const auto batch_idx = src_sampled_nodes[i].first;
+                    sampler.temporal_sample(
+                        /*global_src_node=*/src_sampled_nodes[i],
+                        /*local_src_node=*/i, count, seed_times[batch_idx],
+                        dst_time_data, dst_mapper, generator,
+                        dst_sampled_nodes);
+                  }
+                }
+              }
+            }
+          });
+
+      if (parallel) {
+        for (auto& dst_sampled : dst_sampled_nodes_dict) {
+          std::copy(dst_sampled.second.begin(), dst_sampled.second.end(),
+                    std::back_inserter(sampled_nodes_dict[dst_sampled.first]));
+        }
       }
+      at::parallel_for(0, node_types.size(), 1, [&](size_t _s, size_t _e) {
```

Comment:
Does this bring any gain? We are not doing heavy work here, so the overhead of threading might not be worth it?

Reply:
But maybe to be sure, I'll take 10 measurements and average the results

Reply:
The average of 10 measurements is as follows:
<img width="781" alt="Screenshot 2023-05-03 at 15 38 45" src="https://user-images.githubusercontent.com/58218729/235933020-275820e9-735c-4dd5-b21b-dc5cc64d1b66.png">

Looks like the results are almost the same. Should I remove it then? WDYT?

Reply:
I would be in favor of removing it, as it doesn't look that useful.

Reply:
Sure